## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [368]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [369]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [370]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [371]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [372]:
# What three item categories had the most entries?

top_three = df["category"].value_counts()
top_three.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [373]:
# For the category with the most entries,
# which subcategory had the most entries?
subcat = df["subcategory"].value_counts()
#Used a print statement to test code
#print(subcat)
most_entries_subcat = subcat.idxmax()
#Testing code
#print(most_entries_subcat)
subcat_count = subcat.max()
subcat.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [374]:
# Which five clients had the most entries in the data?

top_five = df["client_id"].value_counts().head(5)
print(top_five)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [375]:
# Store the client ids of those top 5 clients in a list.

top_five_list = top_five.index.tolist()
print(top_five_list)

[33615, 66037, 46820, 38378, 24741]


In [376]:
# How many total units (the qty column) did the
# client with the most entries order order?

most_entries_order = df.loc[(df["client_id"]== 33615),:]
top_total_units = most_entries_order["qty"].sum()
print(top_total_units)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [377]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df["subtotal"] = df["unit_price"] * df["qty"]
df.head(2)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   

   order_week  order_year         item_id     category subcategory  \
0          17        2023  EUD29711-63-6U        decor    wall art   
1          20        2023  XDA18116-89-4A  consumables        pens   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95

In [378]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

df['shipping_price'] = df.apply(lambda row: 7 * row['unit_weight'] * row['qty'] if row['unit_weight'] * row['qty']> 50 else 10 * row['unit_weight'] * row['qty'], axis=1)
df.head(3)



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64

In [379]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
# created a variable for with sales tax converted for multiplication 
sales_tax_rate = .0925 
# variable to store the actual amount of the sales tax
#sales_tax_amt = df['subtotal'] * sales_tax_rate
sales_tax_amt = (df['subtotal'] + df['shipping_price']) * sales_tax_rate
df['total_price'] = df['subtotal'] + df['shipping_price'] + sales_tax_amt

# Rounded to 2 decimal places
df['total_price'] = df['total_price'].round(2)

df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   

   shipping_price  total_price  
0         5512.50    131839.08  
1          312.90       914.26  
2          458.64      1077.12

In [380]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   

   shipping_price  total_price      cost  
0         5512.50    131839.08  85597.05  
1          312.90       914.26    629.79  
2          458.64      1077.12    765.18

In [381]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['cost']
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   

   shipping_price  total_price      cost    profit  
0         5512.50    131839.08  85597.05  46242.03  
1          312.90       914.26    629.79    284.47  
2          458.64      1077.12    765.18    311.94

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [382]:
# Check your work using the totals above
# created a dictionary using order_id as the the key and 
# the total_price as the value
order_total_confirmation = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
# For loop loops through the dict
# Second line of the loop checks for multiple lines matching the order_id in
# the dict and sums up the total_price lines
for order_id, total_price in order_total_confirmation.items():
    confirmed_order = df[df['order_id'] == order_id]['total_price'].sum()
    
    print(f"Order {order_id} total: ${total_price}")
  




Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [383]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

for client_id, qty in top_five.items():
    total_spent = df[df['client_id'] == client_id]['total_price'].sum()
    
    print(f"Client {client_id} ordered {qty} and spent {total_spent: .2f}")


Client 33615 ordered 220 and spent  8377308.52
Client 66037 ordered 211 and spent  10259514.79
Client 46820 ordered 209 and spent  9743794.36
Client 38378 ordered 207 and spent  12906550.87
Client 24741 ordered 207 and spent  82268892.02


In [384]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# My original code reference the top_five as a data frame when it was a series. After 2 weeks of struggle, I ran the error through ChatGPT and finally 
#figured out the issue. ChaptGPT helped me with the code to reference the top_five_list.

top_clients = df[df['client_id'].isin(top_five_list)]

#This code groups by client_id and calculated totals
totals_df = top_clients.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'profit': 'sum'

}).reset_index()

# Code sorts by total profit in descending order
totals_df = totals_df.sort_values(by = 'profit', ascending = False)

print(totals_df)



   client_id     qty  shipping_price  total_price       profit
0      24741  239862      5126448.37  82268892.02  36579992.31
2      38378   73667      3429455.40  12906550.87   3271829.89
4      66037   43018      1395151.85  10259514.79   3255031.81
3      46820   75768      1601448.84   9743794.36   2736602.72
1      33615   64313      1828984.89   8377308.52   2201994.61


In [385]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
# Get a list of column names
# df.columns
new_column_names = ['First Name', 'Last Name', 'Job Title', 'Phone', 'Email', 
                    'Client ID', 'Order ID', 'Order Date', 'Order Week', 
                    'Order Year', 'Item ID', 'Category', 'Subcategory', 'Unit Price',
                    'Unit Cost', 'Unit Weight', 'Quantity', 'Line Number', 'Subtotal', 
                    'Shipping Price', 'Total Price', 'Cost', 'Profit']
df.columns = new_column_names
# Checking new column names
# df.columns
updated_currency = ['Subtotal', 'Shipping Price', 'Total Price', 'Cost', 'Profit']
df[updated_currency] = df[updated_currency] / 1000000

df = df.round(2)

for col in updated_currency:
    df[col] = df[col].apply(lambda x: f"${x:.2f}M")



df.head(10)


First Name Last Name                                     Job Title  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 Phone                                Email  Client ID  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   Order ID  Order Date  Order Week  Order Year         Item ID     Category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          Subcategory  Unit Price  Unit Cost  Unit Weight  Quantity  \
0            wall art     1096.80     762.71         7.50       105   
1                pens       24.95      15.09         1.49        21   
2  project management       13.52       7.86         1.68        39   
3                pens       36.42      24.85         1.23        29   
4                misc      195.10     108.17        46.43        20   
5              plants      172.84     154.44         2.79        78   
6    kitchen supplies      245.56     231.95         1.73        56   
7                pens       20.99      16.97         5.12        22   
8      paper products       83.86      59.00         0.27        49   
9             seating       66.88      54.15         2.47        78   

   Line Number Subtotal Shipping Price Total Price    Cost  Profit  
0            1   $0.12M         $0.01M      $0.13M  $0.09M  $0.05M  
1            0   $0.00M         $0.00M      $0.00M  $0.00M  $0.00M  
2            6   $0.00M         $0.00M      $0.00M  $0.00M  $0.00M  
3            3   $0.00M         $0.00M      $0.00M  $0.00M  $0.00M  
4            1   $0.00M         $0.01M      $0.01M  $0.01M  $0.00M  
5            0   $0.01M         $0.00M      $0.02M  $0.01M  $0.00M  
6            5   $0.01M         $0.00M      $0.02M  $0.01M  $0.00M  
7            4   $0.00M         $0.00M      $0.00M  $0.00M  $0.00M  
8            3   $0.00M         $0.00M      $0.00M  $0.00M  $0.00M  
9            1   $0.01M         $0.00M   

In [386]:
# Sort the updated data by "Total Profit" form highest to lowest
df = df.sort_values(by='Profit', ascending = False)
df.head(3)

First Name  Last Name                                 Job Title  \
40161    Charles  Mccormick  Conservation officer, historic buildings   
11444      James     Brooks                        Charity fundraiser   
29737    William     Savage                         Minerals surveyor   

                       Phone                              Email  Client ID  \
40161  001-667-790-6949x2885  mccormick_sign@potter-sanford.com      46571   
11444   001-866-994-6097x017            brooks.j@morgan-fox.com      18789   
29737           534.689.6772          savagewilliam@johnson.com      30339   

       Order ID  Order Date  Order Week  Order Year         Item ID  \
40161   1553623  2023-05-22          21        2023  EMC36732-26-6B   
11444   5037924  2023-05-29          22        2023  MDU78510-11-2X   
29737   9375270  2023-02-14           7        2023  BDC65078-60-7M   

          Category     Subcategory  Unit Price  Unit Cost  Unit Weight  \
40161    furniture         seating       99.45      60.07        23.49   
11444  consumables            misc      407.01     379.07         2.51   
29737  consumables  paper products      229.74     198.38        27.51   

       Quantity  Line Number Subtotal Shipping Price Total Price     Cost  \
40161    127184            6  $12.65M        $20.91M     $36.67M  $28.55M   
11444    109328            0  $44.50M         $1.92M     $50.71M  $43.36M   
29737    100479            7  $23.08M        $19.35M     $46.36M  $39.28M   

       Profit  
40161  $8.11M  
11444  $7.35M  
29737  $7.08M

In [387]:
summary_findings = "The most profitable category and subcategory were Furniture and Seating. The least profitable category and subcategory were Decor and Wall Art. The top client spent $82268892.02."

print(summary_findings)

The most profitable category and subcategory were Furniture and Seating. The least profitable category and subcategory were Decor and Wall Art. The top client spent $82268892.02.
